sqlalchemy_bigquery/base.py

```diff
@@ -830,23 +832,61 @@ def _process_time_partitioning(
         - Given a table with a TIMESTAMP type column 'event_timestamp' and setting
         'time_partitioning.field' to 'event_timestamp', the function returns
         "PARTITION BY TIMESTAMP_TRUNC(event_timestamp, DAY)".
+
+        Current inputs allowed by BQ and covered by this function include:
+        * _PARTITIONDATE
+        * DATETIME_TRUNC(<datetime_column>, DAY/HOUR/MONTH/YEAR)
+        * TIMESTAMP_TRUNC(<timestamp_column>, DAY/HOUR/MONTH/YEAR)
+        * DATE_TRUNC(<date_column>, MONTH/YEAR)
+
+        Additional allowed options not explicitly covered by this function
+        include:
+        * DATE(_PARTITIONTIME)
+        * DATE(<timestamp_column>)
+        * DATE(<datetime_column>)
+        * DATE column
         """
+
         field = "_PARTITIONDATE"
         trunc_fn = "DATE_TRUNC"
 
+        # Format used with _PARTITIONDATE which can only be used for
```

Comment:
There are a lot of cases here that make this function hard to parse. In particular, I think it could be improved if instead of the first special casing for time.partitioning.field is None block was instead combined by deleting that check, and moving the if time.partioning.type_ in ... below the if field is not None block:

```python
        field = "_PARTITIONDATE"
        trunc_fn = "DATE_TRUNC"
        if time_partitioning.field is not None:
            field = time_partitioning.field

            if isinstance(
                table.columns[field].type,
                (sqlalchemy.sql.sqltypes.TIMESTAMP),
            ):
                trunc_fn = "TIMESTAMP_TRUNC"
            elif isinstance(
                table.columns[field].type,
                sqlalchemy.sql.sqltypes.DATETIME,
            ):
                  trunc_fn = "DATETIME_TRUNC"



        if trunc_fn == "DATE_TRUNC" and not time_partitioning.type_ in {"DAY", "MONTH", "YEAR"}:
               raise ValueError(
                      f"DATE_TRUNC can only be used with TimePartitioningTypes {{DAY, MONTH, YEAR}} received {time_partitioning.type_}"
                  )

        # Format used with generically with DATE, TIMESTAMP, DATETIME, DATE_TRUNC
        return f"PARTITION BY {trunc_fn}({field}, {time_partitioning.type_})"**
```

Reply:
I had removed the the last `if isinstance(table.columns[field].type,`, because I thought it was part of the if elif, I see now that its separate. If thats the case, it could be the first if condition so it can take priority instead of setting it back after the fact (pattern matching would be nice here when we support >= python 3.10)

Reply:
@suzmue 

I am considering an approach such as this. It separates each decision point into concrete steps and incrementally leads to a final choice.

It should be easily expandable later when/if we add additional conditions (i.e. _PARTITIONTIME or the use of the DATE() function with specific column types).

Using a mapping such as `sqltypes` also puts all the elements of the truth table in one concise spot (i.e. if using TIMESTAMP, the function will be TIMESTAMP_TRUNC and the allowable partition_periods are D/H/M/Y) which is the important part and allows us to have very simple test conditions and dictionary lookup in the code.

NOTE: I have not tested this. Consider this pseudocode while we talk through the concept.
 
```
from sqlalchemy.sql.sqltypes import (
    TIMESTAMP,
    DATETIME,
    DATE,
)

sqltypes = {
    _PARTITIONDATE: ("_PARTITIONDATE", None),
    TIMESTAMP: ("TIMESTAMP_TRUNC", {DAY, HOUR, MONTH, YEAR}),
    DATETIME: ("DATETIME_TRUNC", {DAY, HOUR, MONTH, YEAR}),
    DATE: ("DATE_TRUNC", {DAY, MONTH, YEAR}),
}

# Extract field (i.e <column_name> or _PARTITIONDATE)
# AND extract the column_type (i.e. TIMESTAMP, DATE, DATETIME, _PARTITIONDATE)
if time_partitioning.field is not None:
    field = time_partitioning.field
    column_type = table.columns[field].type

else:
    field = "_PARTITIONDATE"
    column_type = _PARTITIONDATE

# Extract time_partitioning.type_ (DAY, HOUR, MONTH, YEAR)
# i.e. generates one partition per type (1/DAY, 1/HOUR)
if time_partitioning.type_ is not None:
    partitioning_period = time_partitioning.type_

# Extract the truncation_function (i.e. DATE_TRUNC)
# and the set of allowable partition_periods 
# that can be used in that function
trunc_fn, allowed_partitions = sqltypes[column_type]

# Create output:

# Special case: _PARTITIONDATE does NOT use a function or partitioning_period
if trunc_fn == "_PARTITIONDATE":
    return f"PARTITION BY {field}"

# Generic case 
if partitioning_period not in allowed_partitions:
    raise ValueError("error msg")

return f"PARTITION BY {trunc_fn}({field}), {partitioning_period}"
```

Reply:
Your approach lgtm!